data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -189,52 +201,59 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
         if (forceFlush || ThresholdCheck.checkThresholdExceed(currentBufferPerBatch, maxEvents, maxBytes, maxCollectionDuration)) {
             try {
                 requestCodec.complete(currentBufferPerBatch.getOutputStream());
 
                 // Capture buffer before resetting
-                final Buffer flushedBuffer = currentBufferPerBatch;
                 final int eventCount = currentBufferPerBatch.getEventCount();
 
-                CompletableFuture<InvokeResponse> future = flushedBuffer.flushToLambda(invocationType);
+                CompletableFuture<InvokeResponse> future = currentBufferPerBatch.flushToLambda(invocationType);
 
                 // Handle future
                 CompletableFuture<Void> processingFuture = future.thenAccept(response -> {
                     //Success handler
-                    handleLambdaResponse(resultRecords, flushedBuffer, eventCount, response);
+                    handleLambdaResponse(resultRecords, currentBufferPerBatch, eventCount, response, responseCodec);
                 }).exceptionally(throwable -> {
                     //Failure handler
-                    LOG.error(NOISY, "Exception occurred while invoking Lambda. Function: {}, event in batch:{} | Exception: ", functionName, currentBufferPerBatch.getRecords().get(0), throwable);
-                    requestPayloadMetric.set(flushedBuffer.getPayloadRequestSize());
+                    List<Record<Event>> bufferRecords = currentBufferPerBatch.getRecords();
+                    Record<Event> eventRecord = bufferRecords.isEmpty() ? null : bufferRecords.get(0);
+                    LOG.error(NOISY, "Exception occurred while invoking Lambda. Function: {} , Event: {}",
+                                functionName, eventRecord == null? "null":eventRecord.getData(), throwable);
+                    requestPayloadMetric.set(currentBufferPerBatch.getPayloadRequestSize());
                     responsePayloadMetric.set(0);
-                    Duration latency = flushedBuffer.stopLatencyWatch();
+                    Duration latency = currentBufferPerBatch.stopLatencyWatch();
                     lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
-                    handleFailure(throwable, flushedBuffer, resultRecords);
+                    handleFailure(throwable, currentBufferPerBatch, resultRecords);
                     return null;
                 });
 
                 futureList.add(processingFuture);
-
-                // Create a new buffer for the next batch
-                currentBufferPerBatch = lambdaCommonHandler.createBuffer(currentBufferPerBatch);
             } catch (IOException e) {
                 LOG.error(NOISY, "Exception while flushing to lambda", e);
                 handleFailure(e, currentBufferPerBatch, resultRecords);
-                currentBufferPerBatch = lambdaCommonHandler.createBuffer(currentBufferPerBatch);
+            } finally {
+                return true;
```

Comment:
I'm not sure it makes sense to have a `return` in a `finally` block. I think you should just have `return true` after the block.